pages/blog/graphql-error-handling-with-fp.mdx

```diff
@@ -0,0 +1,640 @@
+
+```graphql
+interface BaseError {
+  message: String!
+}
+
+type InvalidInputError implements BaseError {
+  message: String!
+}
+
+type NotFoundError implements BaseError {
+  message: String!
+}
+
+type UnknownError implements BaseError {
+  message: String!
+}
+
+type NotAllowedError implements BaseError {
+  message: String!
+}
+
+type Entity {
+  id: ID!
+  name: String!
+}
+
+union EntityResult =
+  | Entity
+  | NotFoundError
+  | NotAllowedError
+  | InvalidInputError
+  | UnknownError
+
+type Query {
+  entity(id: ID!, userId: ID!): EntityResult!
+}
+```
+
+We define a `BaseError` interface that all concrete Errors implement.
+Then, we define the query result as a Union type of the happy path (an `Entity`) and the other possible results (`NotFoundError`, `NotAllowedError`, etc.). In this manner, we extensively describe all possible results, with the added benefit that we can make our result type non-nullable (with the `!` character).
+
+For a more detailed explanation, you can read this article from Sasha Solomon: [200 OK! Error Handling in GraphQL](https://sachee.medium.com/200-ok-error-handling-in-graphql-7ec869aec9bc).
+
+## Handle those Schema changes into the GraphQL server
+
+Since we changed the Schema, we also need to change the query, mutation and types resolvers.
+
+For simple cases, we can just return the correct object from the query resolver:
+
+```typescript
+const resolvers = {
+  Query: {
+    entity: async (parent, { id, userId }, context) => {
+      const entity = await context.api.getEntityById(id);
+      const isAllowed = await context.api.isUserAllowedForEntity(id, userId);
+
+      if (!isAllowed) {
+        return {
+          __typename: 'NotAllowedError',
+          message: `User ${userId} not allowed for entity ${id}`,
+        };
+      }
+      if (!entity) {
+        return {
+          __typename: 'NotFoundError',
+          message: `Entity ${id} not found`,
+        };
+      }
+      return {
+        __typename: 'Entity',
+        ...entity,
+      };
+    },
+  },
+};
+```
+
+Since we include the `__typename` and the fields match the Schema type definitions, there is no need to define type resolvers.
+
+For more details on this solution, you can read this article from [Laurin Quast](https://github.com/n1ru4l): [Handling GraphQL errors like a champ with unions and interfaces](https://blog.logrocket.com/handling-graphql-errors-like-a-champ-with-unions-and-interfaces/).
+
+
+# Typesafe error handling on the client-side
+
+
+Let's now take it to the next level using Typescript and GraphQL Code Generator.
+
+[Typescript](https://www.typescriptlang.org/) is a superset of Javascript which adds type annotations to Javascript code and therefore type safety (with caveats) at compile-time.
+
+[GraphQL Code Generator](https://www.graphql-code-generator.com/) is a tool which generates type definitions which corresponds to the GraphQL schema. It has several plugins and we'll use 3 of those:
```

Comment:
I would add here "developed by [The Guild](https://the-guild.dev/)"

Reply:
> What a great post, thanks @gthau !

And thank you for your suggestions ;)

> I would for this post to be the thing we would link to whenever people are asking us about error handling in GraphQL. We have a Notion doc where we keep track of our thoughts around that, do you think you can go through that and see if you can/should cover everything in there also here and we would deprecate that doc in favor of your post? https://theguildoss.notion.site/Error-handling-in-GraphQL-343aba136d3f4228a7d798b88bde2165

I'll go through it and check what can be incorporated.
 
> Also on the subject of creating types for the results of remote APIs/DBs, what are your thought about generating these types from schemas and things like that? Is that worth mentioning here?

Indeed usually the types will be auto-generated from Graphql schema, JSON schema for REST API or from database schema. For the purpose of the article, I wanted to concentrate on the typed error handling and the safety and API self-documentation the monads bring, so I tried to make the repo the bare minimum to showcase this.

Not sure if I should use a real API in the showcase repo to demo the remote APIs type generation. I don't know if it risks diluting the intent of the article.

Reply:
yes I agree about the type generation. maybe worth mentioning that option in a single sentence? with links to further reading?

Reply:
> yes I agree about the type generation. maybe worth mentioning that option in a single sentence? with links to further reading?

Ok, will do.

Reply:
> Also on the subject of creating types for the results of remote APIs/DBs, what are your thought about generating these types from schemas and things like that?
> Is that worth mentioning here?

Added a note about it, see commit https://github.com/the-guild-org/the-guild-website/pull/611/commits/a2787c37f6de45cea720cf0865259520583c3e77